`experimetal_test_shell_command` cannot find test script

I was trying to use `experimetal_test_shell_command` to run a shell-based test using the latest Pants `main` branch, but the `test.sh` script is not found.

The BUILD file in question is https://github.com/tdyas/pants-go-testing/blob/main/race/BUILD.

I ran (using scie-pants but also occurs with pants_from_sources script):

```
% PANTS_SOURCE=../pants pants test race::
11:43:02.85 [INFO] Completed: Running the `experimental_test_shell_command` at `race:racy_test`
11:43:02.85 [ERROR] Completed: Test with shell command - race:racy_test - failed (exit code 127).
race:racy_test: ./test.sh: No such file or directory



✕ race:racy_test failed in 0.01s.
```

--- a/src/python/pants/backend/shell/register.py
+++ b/src/python/pants/backend/shell/register.py
@@ -15,7 +15,7 @@ from pants.backend.shell.target_types import (
     Shunit2TestTarget,
 )
 from pants.backend.shell.target_types import rules as target_types_rules
-from pants.backend.shell.util_rules import shell_command
+from pants.backend.shell.util_rules import run_in_sandbox, shell_command
 
 
 def target_types():
@@ -35,6 +35,7 @@ def rules():
     return [
         *dependency_inference.rules(),
         *shell_command.rules(),
+        *run_in_sandbox.rules(),
         *shunit2.rules(),
         *shunit2_test_runner.rules(),
         *tailor.rules(),

--- a/src/python/pants/backend/shell/target_types.py
+++ b/src/python/pants/backend/shell/target_types.py
@@ -278,8 +278,7 @@ class ShellCommandOutputsField(StringSequenceField):
     alias = "outputs"
     help = softwrap(
         """
-        Specify the shell command output files and directories, relative to the `BUILD` file's
-        directory.
+        Specify the shell command output files and directories, relative to the value of `workdir`.
 
         Use a trailing slash on directory names, i.e. `my_dir/`.
 
@@ -297,8 +296,7 @@ class ShellCommandOutputFilesField(StringSequenceField):
     default = ()
     help = softwrap(
         """
-        Specify the shell command's output files to capture, relative to the `BUILD` file's
-        directory.
+        Specify the shell command's output files to capture, relative to the value of `workdir`.
 
         For directories, use `output_directories`. At least one of `output_files` and
         `output_directories` must be specified.
@@ -316,7 +314,7 @@ class ShellCommandOutputDirectoriesField(StringSequenceField):
     help = softwrap(
         """
         Specify full directories (including recursive descendants) of output to capture from the
-        shell command, relative to the `BUILD` file's directory.
+        shell command, relative to the value of `workdir`.
 
         For individual files, use `output_files`. At least one of `output_files` and
         `output_directories` must be specified.

--- a/src/python/pants/backend/shell/target_types.py
+++ b/src/python/pants/backend/shell/target_types.py
@@ -430,15 +430,21 @@ class ShellCommandLogOutputField(BoolField):
 
 class ShellCommandWorkdirField(StringField):
     alias = "workdir"
-    default = None
+    default = "."
     help = softwrap(
-        "Sets the current working directory of the command, relative to the project root. If not "
-        "set, use the project root.\n\n"
+        "Sets the current working directory of the command. "
         "To specify the location of the `BUILD` file, use `.`. Values beginning with `.` are "
         "relative to the location of the `BUILD` file.\n\n"
-        "To specify the project/build root, use `/` or the empty string."
+        "To specify the build root, use `/` or the empty string.\n\n"
+        "Values that do not begin with `.` or `/` are relative to the build root."
     )
 
+class RunShellCommandWorkdirField(ShellCommandWorkdirField):
+    default = None
+    help = softwrap(
+        "Sets the current working directory of the command that is `run`. If `None`, run the "
+        "command from the current working directory."
+    )
 
 class ShellCommandTestDependenciesField(ShellCommandExecutionDependenciesField):
     pass
@@ -536,7 +542,7 @@ class ShellCommandRunTarget(Target):
         *COMMON_TARGET_FIELDS,
         ShellCommandExecutionDependenciesField,
         ShellCommandCommandField,
-        ShellCommandWorkdirField,
+        RunShellCommandWorkdirField,
         ShellCommandIsInteractiveField,
     )
     help = softwrap(

--- a/src/python/pants/backend/shell/target_types.py
+++ b/src/python/pants/backend/shell/target_types.py
@@ -5,6 +5,7 @@ from __future__ import annotations
 
 import re
 from enum import Enum
+from typing import ClassVar, Optional
 
 from pants.backend.shell.subsystems.shell_setup import ShellSetup
 from pants.core.goals.test import RuntimePackageDependenciesField, TestTimeoutField
@@ -430,7 +431,7 @@ class ShellCommandLogOutputField(BoolField):
 
 class ShellCommandWorkdirField(StringField):
     alias = "workdir"
-    default = "."
+    default: ClassVar[Optional[str]] = "."
     help = softwrap(
         "Sets the current working directory of the command. "
         "To specify the location of the `BUILD` file, use `.`. Values beginning with `.` are "
@@ -439,12 +440,27 @@ class ShellCommandWorkdirField(StringField):
         "Values that do not begin with `.` or `/` are relative to the build root."
     )
 
+
 class RunShellCommandWorkdirField(ShellCommandWorkdirField):
     default = None
     help = softwrap(
         "Sets the current working directory of the command that is `run`. If `None`, run the "
         "command from the current working directory."
     )
+
+
+class ShellCommandOutputRootDirField(StringField):
+    alias = "root_output_directory"
+    default = "."
+    help = softwrap(
+        "Adjusts the location of files output by this command, when consumed as a dependency.\n\n"
+        "`.` or values beginning with `./` are relative to the location of the `BUILD` file.\n\n"
+        "`.` or values beginning with `./` are relative to the value of `workdir`.\n\n"
+        "To specify the build root, use `/` or the empty string.\n\n"
+        "Values that do not begin with `.` or `/` are relative to the build root.\n\n"
+        "All files output by "
+    )
+
 
 class ShellCommandTestDependenciesField(ShellCommandExecutionDependenciesField):
     pass
@@ -472,6 +488,7 @@ class ShellCommandTarget(Target):
         ShellCommandToolsField,
         ShellCommandExtraEnvVarsField,
         ShellCommandWorkdirField,
+        ShellCommandOutputRootDirField,
         EnvironmentField,
     )
     help = softwrap(
@@ -515,6 +532,7 @@ class ShellRunInSandboxTarget(Target):
         ShellCommandToolsField,
         ShellCommandExtraEnvVarsField,
         ShellCommandWorkdirField,
+        ShellCommandOutputRootDirField,
         EnvironmentField,
     )
     help = softwrap(

--- a/src/python/pants/backend/shell/target_types.py
+++ b/src/python/pants/backend/shell/target_types.py
@@ -431,11 +431,12 @@ class ShellCommandWorkdirField(StringField):
     alias = "workdir"
     default: ClassVar[Optional[str]] = "."
     help = softwrap(
-        "Sets the current working directory of the command. "
-        "To specify the location of the `BUILD` file, use `.`. Values beginning with `.` are "
-        "relative to the location of the `BUILD` file.\n\n"
-        "To specify the build root, use `/` or the empty string.\n\n"
-        "Values that do not begin with `.` or `/` are relative to the build root."
+        "Sets the current working directory of the command. \n\n"
+        "Values are relative to the build root, except in the following cases:\n\n"
+        "* `.` specifies the location of the `BUILD` file.\n"
+        "* Values beginning with `./` are relative to the location of the `BUILD` file.\n"
+        "* `/` or the empty string specifies the build root.\n"
+        "* Values beginning with `/` are also relative to the build root."
     )
 
 
@@ -443,7 +444,7 @@ class RunShellCommandWorkdirField(ShellCommandWorkdirField):
     default = None
     help = softwrap(
         "Sets the current working directory of the command that is `run`. If `None`, run the "
-        "command from the current working directory."
+        "command from the directory you are invoking Pants from."
     )
 
 

--- a/src/python/pants/backend/shell/util_rules/adhoc_process_support.py
+++ b/src/python/pants/backend/shell/util_rules/adhoc_process_support.py
@@ -10,6 +10,7 @@ import re
 import shlex
 from dataclasses import dataclass
 from textwrap import dedent  # noqa: PNT20
+from typing import Union
 
 from pants.backend.shell.subsystems.shell_setup import ShellSetup
 from pants.backend.shell.target_types import (
@@ -34,6 +35,7 @@ from pants.core.util_rules.system_binaries import (
 from pants.engine.addresses import Addresses, UnparsedAddressInputs
 from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
 from pants.engine.fs import EMPTY_DIGEST, CreateDigest, Digest, Directory, MergeDigests, Snapshot
+from pants.engine.internals.native_engine import RemovePrefix
 from pants.engine.process import Process
 from pants.engine.rules import Get, MultiGet, collect_rules, rule, rule_helper
 from pants.engine.target import (
@@ -160,6 +162,20 @@ def _parse_outputs_from_command(
         output_files = tuple(f for f in outputs if not f.endswith("/"))
         output_directories = tuple(d for d in outputs if d.endswith("/"))
     return output_files, output_directories
+
+
+@rule_helper
+async def _adjust_root_output_directory(
+    digest: Digest, address: Address, working_directory: str, root_output_directory: str
+) -> Digest:
+
+    working_directory = _parse_working_directory(working_directory, address)
+    new_root = _parse_working_directory(root_output_directory, working_directory)
+
+    if new_root == "":
+        return digest
+    else:
+        return await Get(Digest, RemovePrefix(digest, new_root))
 
 
 def _shell_tool_safe_env_name(tool_name: str) -> str:
@@ -315,10 +331,13 @@ def _output_at_build_root(process: Process, bash: BashBinary) -> Process:
     )
 
 
-def _parse_working_directory(workdir_in: str, address: Address) -> str:
+def _parse_working_directory(workdir_in: str, relative_to: Union[Address, str]) -> str:
     """Convert the `workdir` field into something that can be understood by `Process`."""
 
-    reldir = address.spec_path
+    if isinstance(relative_to, Address):
+        reldir = relative_to.spec_path
+    else:
+        reldir = relative_to
 
     if workdir_in == ".":
         return reldir

--- a/src/python/pants/backend/shell/util_rules/run_in_sandbox.py
+++ b/src/python/pants/backend/shell/util_rules/run_in_sandbox.py
@@ -133,7 +133,10 @@ async def run_in_sandbox_request(
             logger.warning(result.stderr.decode())
 
     adjusted = await _adjust_root_output_directory(
-        result.output_digest, shell_command.address, root_output_directory, working_directory
+        result.output_digest,
+        shell_command.address,
+        working_directory,
+        root_output_directory,
     )
     output = await Get(Snapshot, Digest, adjusted)
 

--- a/src/python/pants/backend/shell/util_rules/run_in_sandbox.py
+++ b/src/python/pants/backend/shell/util_rules/run_in_sandbox.py
@@ -10,10 +10,12 @@ from pants.backend.shell.target_types import (
     RunInSandboxRunnableField,
     RunInSandboxSourcesField,
     ShellCommandLogOutputField,
+    ShellCommandOutputRootDirField,
     ShellCommandWorkdirField,
 )
 from pants.backend.shell.util_rules.adhoc_process_support import (
     ShellCommandProcessRequest,
+    _adjust_root_output_directory,
     _execution_environment_from_dependencies,
     _parse_outputs_from_command,
 )
@@ -82,6 +84,7 @@ async def run_in_sandbox_request(
     run_field_set: RunFieldSet = field_sets.field_sets[0]
 
     working_directory = shell_command[ShellCommandWorkdirField].value or ""
+    root_output_directory = shell_command[ShellCommandOutputRootDirField].value or ""
 
     # Must be run in target environment so that the binaries/envvars match the execution
     # environment when we actually run the process.
@@ -129,7 +132,11 @@ async def run_in_sandbox_request(
         if result.stderr:
             logger.warning(result.stderr.decode())
 
-    output = await Get(Snapshot, Digest, result.output_digest)
+    adjusted = await _adjust_root_output_directory(
+        result.output_digest, shell_command.address, root_output_directory, working_directory
+    )
+    output = await Get(Snapshot, Digest, adjusted)
+
     return GeneratedSources(output)
 
 

--- a/src/python/pants/backend/shell/util_rules/shell_command.py
+++ b/src/python/pants/backend/shell/util_rules/shell_command.py
@@ -154,7 +154,7 @@ async def run_shell_command(
     working_directory = shell_command[ShellCommandWorkdirField].value or ""
     root_output_directory = shell_command[ShellCommandOutputRootDirField].value or ""
     adjusted = await _adjust_root_output_directory(
-        result.output_digest, shell_command.address, root_output_directory, working_directory
+        result.output_digest, shell_command.address, working_directory, root_output_directory
     )
     output = await Get(Snapshot, Digest, adjusted)
     return GeneratedSources(output)

--- a/src/python/pants/backend/shell/util_rules/shell_command.py
+++ b/src/python/pants/backend/shell/util_rules/shell_command.py
@@ -11,6 +11,7 @@ from pants.backend.shell.target_types import (
     ShellCommandExtraEnvVarsField,
     ShellCommandIsInteractiveField,
     ShellCommandLogOutputField,
+    ShellCommandOutputRootDirField,
     ShellCommandSourcesField,
     ShellCommandTimeoutField,
     ShellCommandToolsField,
@@ -18,6 +19,7 @@ from pants.backend.shell.target_types import (
 )
 from pants.backend.shell.util_rules.adhoc_process_support import (
     ShellCommandProcessRequest,
+    _adjust_root_output_directory,
     _execution_environment_from_dependencies,
     _parse_outputs_from_command,
 )
@@ -149,7 +151,12 @@ async def run_shell_command(
         if result.stderr:
             logger.warning(result.stderr.decode())
 
-    output = await Get(Snapshot, Digest, result.output_digest)
+    working_directory = shell_command[ShellCommandWorkdirField].value or ""
+    root_output_directory = shell_command[ShellCommandOutputRootDirField].value or ""
+    adjusted = await _adjust_root_output_directory(
+        result.output_digest, shell_command.address, root_output_directory, working_directory
+    )
+    output = await Get(Snapshot, Digest, adjusted)
     return GeneratedSources(output)
 
 

--- a/src/python/pants/backend/shell/util_rules/shell_command_test.py
+++ b/src/python/pants/backend/shell/util_rules/shell_command_test.py
@@ -133,7 +133,6 @@ def test_sources_and_files(rule_runner: RuleRunner) -> None:
                   output_files=["message.txt"],
                   output_directories=["res"],
                   command="./script.sh",
-                  workdir=".",
                 )
 
                 files(
@@ -161,8 +160,8 @@ def test_sources_and_files(rule_runner: RuleRunner) -> None:
         rule_runner,
         Address("src", target_name="hello"),
         expected_contents={
-            "src/message.txt": RES,
-            "src/res/log.txt": RES,
+            "message.txt": RES,
+            "res/log.txt": RES,
         },
     )
 
@@ -198,8 +197,51 @@ def test_chained_shell_commands(rule_runner: RuleRunner) -> None:
                 experimental_shell_command(
                   name="msg",
                   tools=["echo"],
+                  output_files=["../msg"],
+                  command="echo 'shell_command:a' > ../msg",
+                  root_output_directory="/",
+                )
+                """
+            ),
+            "src/b/BUILD": dedent(
+                """\
+                experimental_shell_command(
+                  name="msg",
+                  tools=["cp", "echo"],
+                  output_files=["../msg"],
+                  command="echo 'shell_command:b' >> ../msg",
+                  execution_dependencies=["src/a:msg"],
+                  root_output_directory="/",
+                )
+                """
+            ),
+        }
+    )
+
+    assert_shell_command_result(
+        rule_runner,
+        Address("src/a", target_name="msg"),
+        expected_contents={"src/msg": "shell_command:a\n"},
+    )
+
+    assert_shell_command_result(
+        rule_runner,
+        Address("src/b", target_name="msg"),
+        expected_contents={"src/msg": "shell_command:a\nshell_command:b\n"},
+    )
+
+
+def test_chained_shell_commands_with_workdir(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/a/BUILD": dedent(
+                """\
+                experimental_shell_command(
+                  name="msg",
+                  tools=["echo"],
                   output_files=["msg"],
                   command="echo 'shell_command:a' > msg",
+                  workdir="/",
                 )
                 """
             ),
@@ -211,6 +253,7 @@ def test_chained_shell_commands(rule_runner: RuleRunner) -> None:
                   output_files=["msg"],
                   command="echo 'shell_command:b' >> msg",
                   execution_dependencies=["src/a:msg"],
+                  workdir="/",
                 )
                 """
             ),
@@ -227,48 +270,6 @@ def test_chained_shell_commands(rule_runner: RuleRunner) -> None:
         rule_runner,
         Address("src/b", target_name="msg"),
         expected_contents={"msg": "shell_command:a\nshell_command:b\n"},
-    )
-
-
-def test_chained_shell_commands_with_workdir(rule_runner: RuleRunner) -> None:
-    rule_runner.write_files(
-        {
-            "src/a/BUILD": dedent(
-                """\
-                experimental_shell_command(
-                  name="msg",
-                  tools=["echo"],
-                  output_files=["msg"],
-                  command="echo 'shell_command:a' > msg",
-                  workdir=".",
-                )
-                """
-            ),
-            "src/b/BUILD": dedent(
-                """\
-                experimental_shell_command(
-                  name="msg",
-                  tools=["cp", "echo"],
-                  output_files=["msg"],
-                  command="cp ../a/msg . ; echo 'shell_command:b' >> msg",
-                  execution_dependencies=["src/a:msg"],
-                  workdir=".",
-                )
-                """
-            ),
-        }
-    )
-
-    assert_shell_command_result(
-        rule_runner,
-        Address("src/a", target_name="msg"),
-        expected_contents={"src/a/msg": "shell_command:a\n"},
-    )
-
-    assert_shell_command_result(
-        rule_runner,
-        Address("src/b", target_name="msg"),
-        expected_contents={"src/b/msg": "shell_command:a\nshell_command:b\n"},
     )
 
 
@@ -390,7 +391,7 @@ def test_package_dependencies(caplog, rule_runner: RuleRunner) -> None:
 
                 experimental_shell_command(
                   name="test",
-                  command="ls src",
+                  command="ls",
                   tools=["ls"],
                   log_output=True,
                   execution_dependencies=[":msg-archive"],
@@ -422,7 +423,8 @@ def test_execution_dependencies(caplog, rule_runner: RuleRunner) -> None:
                 experimental_shell_command(
                   name="a1",
                   command="echo message > msg.txt",
-                  outputs=["msg.txt"],
+                  output_files=["msg.txt"],
+                  workdir="/",
                 )
 
                 experimental_shell_command(
@@ -430,7 +432,8 @@ def test_execution_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     tools=["cat"],
                     command="cat msg.txt > msg2.txt",
                     execution_dependencies=[":a1",],
-                    outputs=["msg2.txt",],
+                    output_files=["msg2.txt",],
+                    workdir="/",
                 )
 
                 # Fails because runtime dependencies are not exported
@@ -440,6 +443,7 @@ def test_execution_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     tools=["cat"],
                     command="cat msg.txt",
                     execution_dependencies=[":a2",],
+                    workdir="/",
                 )
 
                 # Fails because `output_dependencies` are not available at runtime
@@ -449,6 +453,7 @@ def test_execution_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     command="cat msg.txt",
                     execution_dependencies=(),
                     output_dependencies=[":a1"],
+                    workdir="/",
                 )
 
                 # Fails because runtime dependencies are not fetched transitively
@@ -458,6 +463,7 @@ def test_execution_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     tools=["cat"],
                     command="cat msg.txt",
                     output_dependencies=[":a2"],
+                    workdir="/",
                 )
 
                 # Succeeds because `a1` and `a2` are requested directly
@@ -466,7 +472,8 @@ def test_execution_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     tools=["cat"],
                     command="cat msg.txt msg2.txt > output.txt",
                     execution_dependencies=[":a1", ":a2",],
-                    outputs=["output.txt"],
+                    output_files=["output.txt"],
+                    workdir="/",
                 )
 
                 # Succeeds becuase `a1` and `a2` are requested directly and `output_dependencies`
@@ -476,7 +483,8 @@ def test_execution_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     tools=["cat"],
                     command="cat msg.txt msg2.txt > output.txt",
                     output_dependencies=[":a1", ":a2",],
-                    outputs=["output.txt"],
+                    output_files=["output.txt"],
+                    workdir="/",
                 )
                 """
             ),
@@ -519,6 +527,7 @@ def test_old_style_dependencies(caplog, rule_runner: RuleRunner) -> None:
                   name="a1",
                   command="echo message > msg.txt",
                   outputs=["msg.txt"],
+                  workdir="/",\
                 )
 
                 experimental_shell_command(
@@ -527,6 +536,7 @@ def test_old_style_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     command="cat msg.txt > msg2.txt",
                     dependencies=[":a1",],
                     outputs=["msg2.txt",],
+                    workdir="/",
                 )
 
                 experimental_shell_command(
@@ -535,6 +545,7 @@ def test_old_style_dependencies(caplog, rule_runner: RuleRunner) -> None:
                     command="cat msg.txt msg2.txt > output.txt",
                     dependencies=[":a1", ":a2",],
                     outputs=["output.txt"],
+                    workdir="/",
                 )
                 """
             ),
@@ -723,7 +734,7 @@ def test_run_runnable_in_sandbox_with_workdir(rule_runner: RuleRunner) -> None:
         {
             "src/fruitcake.py": dedent(
                 """\
-                f = open("fruitcake.txt", "w")
+                f = open("src/fruitcake.txt", "w")
                 f.write("fruitcake\\n")
                 f.close()
                 """
@@ -738,8 +749,8 @@ def test_run_runnable_in_sandbox_with_workdir(rule_runner: RuleRunner) -> None:
                 experimental_run_in_sandbox(
                   name="run_fruitcake",
                   runnable=":fruitcake",
-                  output_files=["fruitcake.txt"],
-                  workdir=".",
+                  output_files=["src/fruitcake.txt"],
+                  workdir="/",
                 )
                 """
             ),
@@ -763,7 +774,7 @@ def test_relative_directories(rule_runner: RuleRunner) -> None:
                   tools=["echo"],
                   command='echo foosh > ../foosh.txt',
                   output_files=["../foosh.txt"],
-                  workdir=".",
+                  root_output_directory="/",
                 )
                 """
             ),
@@ -787,7 +798,7 @@ def test_relative_directories_2(rule_runner: RuleRunner) -> None:
                   tools=["echo"],
                   command='echo foosh > ../newdir/foosh.txt',
                   output_files=["../newdir/foosh.txt"],
-                  workdir=".",
+                  root_output_directory="/",
                 )
                 """
             ),
@@ -811,7 +822,7 @@ def test_cannot_escape_build_root(rule_runner: RuleRunner) -> None:
                   tools=["echo"],
                   command='echo foosh > ../../invalid.txt',
                   output_files=["../../invalid.txt"],
-                  workdir=".",
+                  root_output_directory="/",
                 )
                 """
             ),
@@ -861,6 +872,7 @@ def test_working_directory_special_values(
                   runnable=":fruitcake",
                   output_files=["fruitcake.txt"],
                   workdir="{workdir}",
+                  root_output_directory="/",
                 )
                 """
             ),

--- a/src/python/pants/core/util_rules/wrap_source_intergration_test.py
+++ b/src/python/pants/core/util_rules/wrap_source_intergration_test.py
@@ -18,6 +18,7 @@ def test_synthesized_python_is_included_in_package() -> None:
                 execution_dependencies=(),
                 output_files=["hello_world.py",],
                 workdir=".",
+                root_output_directory="/",
             )
 
             experimental_wrap_as_python_sources(

